Add deployment modal for model registry

DaoDaoNoCode opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- https://issues.redhat.com/browse/RHOAIENG-123456 -->
JIRA: [RHOAIENG-8396](https://issues.redhat.com/browse/RHOAIENG-8396)

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->

When you deploy a model based on a model registry, the following modal will show:

<img width="1511" alt="Screenshot 2024-08-07 at 1 27 46 PM" src="https://github.com/user-attachments/assets/83ae3c3a-1af2-490f-948b-97565d0c482d">

If anything wrong with the project, the corresponding error message will pop up:

<img width="1511" alt="Screenshot 2024-08-07 at 1 28 32 PM" src="https://github.com/user-attachments/assets/ad57cdea-2bca-4a1e-8892-1569d95f012d">

Otherwise, the current KServe modal or InferenceService modal will pop up depending on the project's serving platform. The name will be pre-filled as `{modelName} - {versionName}`:

<img width="1511" alt="Screenshot 2024-08-07 at 1 29 29 PM" src="https://github.com/user-attachments/assets/32d22f8a-1ecc-4b2c-8c11-cdc62adcdd12">

There is a helper text if the model registry version contains a prompt model format:

<img width="812" alt="Screenshot 2024-08-07 at 1 32 08 PM" src="https://github.com/user-attachments/assets/24bc1830-44d6-45a1-92a9-d6de3c420c2a">

The data connections part will be decided by the model artifact URI.
If there are more than 1 matches, we don't pre-select the data connection:

<img width="1510" alt="Screenshot 2024-08-07 at 1 32 32 PM" src="https://github.com/user-attachments/assets/78f4ca6b-551a-4dad-a102-6470d6a86b6a">

If there is only one match, we pre-select the recommended data connection:

<img width="1510" alt="Screenshot 2024-08-07 at 1 34 19 PM" src="https://github.com/user-attachments/assets/439bff95-7b62-41b6-a443-7349f036ea95">

If there is no match, we pre-select the new data connection field:

<img width="1510" alt="Screenshot 2024-08-07 at 1 35 14 PM" src="https://github.com/user-attachments/assets/cd1a153d-8b6d-4ea5-a6e7-5759f2e204e9">

The path will always be pre-filled based on the URI parsing result:

<img width="806" alt="Screenshot 2024-08-07 at 1 36 58 PM" src="https://github.com/user-attachments/assets/f1ab15e3-07d2-4660-8164-0544c0e6d57c">

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
1. Create project A, B and C
2. For project A, make it an unsupported serving platform (or don't select a serving platform if both platforms are supported)
3. Go to Model registry page, choose one model registry and click on `Deploy` on the version action
4. Choose project A, check the error message
5. For project B and C, make the serving platform selected (or make your cluster only support one platform)
6. Choose project B, check if the name field is auto-filled
7. Check the data connection field is pre-selected as `New data connection` and `path` field is pre-filled
8. Create one matching secret in project B and 2 matching secrets in project C ("matching" means the secret has the same `bucket`, `endpoint` and `region` as the model registry version artifact URI you are going to deploy with)
9. Choose project B again, check the data connection field is pre-selected as `Existing data connection` and the only matching data connection is labeled and pre-selected
10. Choose project C, check the data connection field is pre-selected as `Existing data connection` and both data connections in the dropdown menu are labeled as recommended, but neither is pre-selected

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
Add a unit test to test the util function that shows the error message when choosing the project.

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

yih-wang commented:
Thanks for the work @DaoDaoNoCode! Just a few comments - 
1. Can you stretch the modal width in the first image to the same width as the other modals?
2. Regarding the pre-filled model name, I can't see what happens if a version has been deployed multiple times. The ideal behavior is if the name has already been used within the selected project, add a postfix such as ‘-2’, ‘-3’ to differentiate it.
3. Regarding the helper text to the prompt model format - the helper text will only show up after a model server is selected, and the `model framework` dropdown is available.
4. For data connection, when there is no match and the new data connection is selected by default - a warning is missing, check details in the [mockup](https://www.figma.com/proto/KTsYhANYQqB0NUrXW7wQH1/Model-Registry?page-id=6782%3A296930&node-id=8238-695915&viewport=941%2C-3865%2C0.11&t=28gM6CEeyGJs3Cpp-1&scaling=min-zoom&content-scaling=fixed&starting-point-node-id=8238%3A699060)

mturley commented on frontend/src/pages/modelRegistry/screens/RegisteredModels/useRegisteredModelContext.ts:
```diff
@@ -0,0 +1,37 @@
+import useRegisteredModelById from '~/concepts/modelRegistry/apiHooks/useRegisteredModelById';
+import { ModelVersion } from '~/concepts/modelRegistry/types';
+
+export type RegisteredModelContext = {
```
I wonder if we should use a word other than `Context` here to name `RegisteredModelContext` and `useRegisteredModelContext`, since it's not a React Context. Maybe `RegisteredModelDeployInfo` or something? I guess we have the preexisting `projectContext` prop in the modals so maybe it's fine. Up to you.

Even if we keep the word Context I think it should have the word Deploy in it to indicate its purpose - `RegisteredModelDeployContext` / `useRegisteredModelDeployContext` maybe?

DaoDaoNoCode replied:
I agree with you and I will go with the `RegisteredModelDeployInfo`. Thanks!

mturley commented on frontend/src/pages/modelRegistry/screens/components/DeployRegisteredModelModal.tsx:
```diff
@@ -0,0 +1,112 @@
+import { ModelVersion } from '~/concepts/modelRegistry/types';
+import { ProjectKind } from '~/k8sTypes';
+import useProjectErrorForRegisteredModel from '~/pages/modelRegistry/screens/RegisteredModels/useProjectErrorForRegisteredModel';
+import ProjectSelector from '~/pages/modelServing/screens/projects/InferenceServiceModal/ProjectSelector';
+import { ManageKServeModalWrapper } from '~/pages/modelServing/screens/projects/kServeModal/ManageKServeModal';
+import useServingPlatformStatuses from '~/pages/modelServing/useServingPlatformStatuses';
+import { getProjectModelServingPlatform } from '~/pages/modelServing/screens/projects/utils';
```
There's also a bunch of stuff being imported across `~/pages/modelServing` and `~/pages/modelRegistry` here - maybe we should take a pass and see if any of it is a good candidate to move to `~/concepts/modelServing`....

...ahh, we don't yet have a `~/concepts/modelServing`. Maybe we should open a followup tech tebt issue to create it and lift some of these things out there after MR MVP. I may be stirring up too much refactoring here with these comments 😄.

Yeah ok, for stuff you're importing from `~/pages/modelServing` let's leave them alone for now and I'll open a followup issue, but for new stuff you're introducing we should consider putting stuff in `~/concepts/modelRegistry` if we can do so cleanly. I'll leave it up to your judgement how much of that is within the scope of this PR.

DaoDaoNoCode replied:
Yeah I agree, I will leave this PR as it is for now because I checked the usage in those hooks, and it seems it's cross-page using many enums/const so I think they should be restructured afterward to lift all of them up to the `/concepts` folder in another ticket.

mturley commented on frontend/src/pages/modelServing/screens/projects/InferenceServiceModal/ManageInferenceServiceModal.tsx:
```diff
@@ -33,11 +37,16 @@ type ManageInferenceServiceModalProps = {
 >;
 
-const ManageInferenceServiceModal: React.FC<ManageInferenceServiceModalProps> = ({
+export const ManageInferenceServiceModalWrapper: React.FC<
+  ManageInferenceServiceModalProps & { children: React.ReactNode }
+> = ({
```
Feel free to leave this as-is, but for each of these modals I wonder if you really need the Wrapper components and introduction of `children` - You could instead just add an optional `projectSection` prop that defaults to the default children you pass below and let the DeployRegisteredModelModal pass its own in that overrides it.

DaoDaoNoCode replied:
Good suggestion, will update!

mturley commented on frontend/src/pages/modelServing/screens/projects/kServeModal/ManageKServeModal.tsx:
```diff
@@ -73,12 +77,17 @@ type ManageKServeModalProps = {
 >;
 
-const ManageKServeModal: React.FC<ManageKServeModalProps> = ({
+export const ManageKServeModalWrapper: React.FC<
```
Same here, you could avoid the wrapper if you want to.

mturley reviewed (changes requested):
This is looking really good! I have one actual change request (the comment in `useLabeledDataConnections`) and then just some questions and suggestions.

DaoDaoNoCode commented on frontend/src/pages/modelRegistry/screens/RegisteredModels/useRegisteredModelContext.ts:
```diff
@@ -0,0 +1,37 @@
+};
+
+const useRegisteredModelContext = (modelVersion: ModelVersion): RegisteredModelContext => {
+  const [registeredModel] = useRegisteredModelById(modelVersion.registeredModelId);
+  const [modelArtifactList] = useModelArtifactsByVersionId(modelVersion.id);
```
Hmm, I am not sure, the only side-effect here is if there is an error, the modal will not pre-fill those values, but that doesn't block users from deploying the model, they just need to select by themselves. I am not sure the best way to handle the loading and error state here, WDYT?

mturley replied:
I suppose the user can't get to this deployment action unless they've already gone through the Model/Version details pages which will be making these same API requests, so it's not critical to catch errors here - but technically if the user e.g. loses their internet connection and then clicks Deploy, we should ideally show an error instead of failing silently.

Can we add two more properties to `RegisteredModelDeployInfo`:
```ts
loaded: boolean;
loadError?: Error;
```
and where we call `useRegisteredModelById` and `useModelArtifactsByVersionId`, can we pull out the `loaded` and `loadError` values from those returned arrays, and then when we return `RegisteredModelDeployInfo` we can aggregate them like:
```ts
loaded: registeredModelLoaded && modelArtifactsLoaded,
loadError: registeredModelLoadError || modelArtifactsLoadError,
```
and then in the `DeployRegisteredModelModal`, in the body of the modal can we render an Alert if there is a `loadError` in that object or a Spinner if it's `!loaded`?

DaoDaoNoCode commented on frontend/src/concepts/modelRegistry/context/ModelRegistryContext.tsx:
```diff
@@ -49,6 +64,14 @@ export const ModelRegistryContextProvider = conditionalArea<ModelRegistryContext
     setDisableTimeout(true);
   }, []);
 
+  const servingRuntimeTemplates = useTemplates(dashboardNamespace);
+  const servingRuntimeTemplateOrder = useContextResourceData<string>(
+    useTemplateOrder(dashboardNamespace),
+  );
+  const servingRuntimeTemplateDisablement = useContextResourceData<string>(
+    useTemplateDisablement(dashboardNamespace),
+  );
+
```
I think it's OK not to capture the state here, the submit function should be able to catch if there is an error when submitting, the error will only affect the order/disablement status of the templates, which is not a big deal.

mturley replied:
Sounds reasonable to me.

DaoDaoNoCode commented on frontend/src/pages/modelServing/screens/projects/kServeModal/ManageKServeModal.tsx:
```diff
@@ -116,6 +131,11 @@ const ManageKServeModal: React.FC<ManageKServeModalProps> = ({
     }
   }, [currentProjectName, setCreateDataInferenceService, isOpen]);
 
+  // Refresh model format selection when changing serving runtime template selection
+  React.useEffect(() => {
+    setCreateDataInferenceService('format', { name: '' });
+  }, [createDataServingRuntime.servingRuntimeTemplateName, setCreateDataInferenceService]);
+
```
Fixed a small bug here. When selecting a new template, it will reset the model frame selection now (it didn't before). Not sure if there is a ticket logged.

mturley replied:
I can't find a ticket, I think it's probably ok to just fix this here. @lucferbux do you happen to know if there's a bug logged for this?

DaoDaoNoCode commented on frontend/src/pages/modelRegistry/screens/RegisteredModels/usePrefillDeployModalFromModelRegistry.ts:
```diff
@@ -0,0 +1,87 @@
+          type: InferenceServiceStorageType.EXISTING_STORAGE,
+        });
+      } else {
+        const prefilledAWSData = [
+          { key: AwsKeys.NAME, value: registeredModelDeployInfo.modelArtifactStorageKey || '' },
+          { key: AwsKeys.AWS_S3_BUCKET, value: storageFields.bucket },
+          { key: AwsKeys.S3_ENDPOINT, value: storageFields.endpoint },
+          { key: AwsKeys.DEFAULT_REGION, value: storageFields.region || '' },
+          ...EMPTY_AWS_SECRET_DATA,
+        ];
```
Pre-filled all the fields if the URI is parsed correctly. Forgot to do that before, and now it's fixed.

DaoDaoNoCode commented:
> 1. Can you stretch the modal width in the first image to the same width as the other modals?
> 2. Regarding the pre-filled model name, I can't see what happens if a version has been deployed multiple times. The ideal behavior is if the name has already been used within the selected project, add a postfix such as ‘-2’, ‘-3’ to differentiate it.
> 3. Regarding the helper text to the prompt model format - the helper text will only show up after a model server is selected, and the model framework dropdown is available.
> 4. For data connection, when there is no match and the new data connection is selected by default - a warning is missing, check details in the [mockup](https://www.figma.com/proto/KTsYhANYQqB0NUrXW7wQH1/Model-Registry?page-id=6782%3A296930&node-id=8238-695915&viewport=941%2C-3865%2C0.11&t=28gM6CEeyGJs3Cpp-1&scaling=min-zoom&content-scaling=fixed&starting-point-node-id=8238%3A699060)

Based on the comments above, I updated the following:
1. For your first comment, I updated the size of the initial modal

<img width="1512" alt="Screenshot 2024-08-08 at 7 48 23 PM" src="https://github.com/user-attachments/assets/2abebed3-1ac1-4ec2-b15a-b89bef261e5b">

2. For the second comment, I found it's a little bit hard from the technical side to implement what you requested. We will need to fetch all the models in that project and it's hard to decide the postfix number (e.g. There are model-1, model-3 and model-4, what should the auto-generated number be? model-2 or model-5?). I prefer we keep it as it is now, and if the name is duplicated, the server will always prompt an error when the user submits the model. (cc @mturley Might need your input here)
3. For the third comment, I updated the helper text to make it only show when there are items in the dropdown menu.

No items in the menu:

<img width="814" alt="Screenshot 2024-08-08 at 7 56 50 PM" src="https://github.com/user-attachments/assets/e3ebf251-0ca4-4672-877e-bb2538a50304">

Items in the menu:

<img width="814" alt="Screenshot 2024-08-08 at 7 57 01 PM" src="https://github.com/user-attachments/assets/7c42ab7d-d1c2-4642-8743-89b4c02d5d71">

4. For the fourth comment, I updated the alert:

<img width="1512" alt="Screenshot 2024-08-08 at 7 58 13 PM" src="https://github.com/user-attachments/assets/d9750ee5-af1d-4a4a-b3a5-367bd9a46f43">

@yih-wang Let me know what you think, thanks!

yih-wang commented:
@DaoDaoNoCode All the changes look good. Thanks for the rapid work!
Regarding the auto-generated deployment name, my primary concern is to avoid duplicate names and prevent user frustration from seeing duplicated names error continuously. If the original proposal doesn't work, we can implement a workaround by adding a timestamp as postfix to each name. So we'll have a format like [modelname-versionname-yyyymmddhhmmss], which should effectively eliminate duplicates as well.

yih-wang commented:
LGTM, thanks @DaoDaoNoCode !

mturley reviewed (approved):
Pulled and tested on the modelregistry-ui cluster - everything looks and works great here!! Thanks so much for the hard work @DaoDaoNoCode !
